File: app/src/main/java/org/tasks/backup/TasksJsonExporter.java
package org.tasks.backup;

import static org.tasks.date.DateTimeUtils.newDateTime;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.net.Uri;
import android.os.Handler;
import android.widget.Toast;
import androidx.annotation.Nullable;
import com.google.common.io.Files;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.todoroo.andlib.utility.DialogUtilities;
import com.todoroo.astrid.backup.BackupConstants;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import java.io.IOException;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.inject.Inject;
import org.tasks.BuildConfig;
import org.tasks.R;
import org.tasks.data.AlarmDao;
import org.tasks.data.CaldavDao;
import org.tasks.data.FilterDao;
import org.tasks.data.GoogleTaskDao;
import org.tasks.data.GoogleTaskListDao;
import org.tasks.data.LocationDao;
import org.tasks.data.TagDao;
import org.tasks.data.TagDataDao;
import org.tasks.data.TaskAttachmentDao;
import org.tasks.data.UserActivityDao;
import org.tasks.drive.DriveInvoker;
import org.tasks.files.FileHelper;
import org.tasks.jobs.WorkManager;
import org.tasks.preferences.Preferences;
import timber.log.Timber;

public class TasksJsonExporter {

  static final Charset UTF_8 = Charset.forName("UTF-8");
  private static final String MIME = "application/json";
  private static final String EXTENSION = ".json";

  // --- public interface
  private final TagDataDao tagDataDao;

  // --- implementation
  private final AlarmDao alarmDao;
  private final LocationDao locationDao;
  private final TagDao tagDao;
  private final GoogleTaskDao googleTaskDao;
  private final FilterDao filterDao;
  private final GoogleTaskListDao googleTaskListDao;
  private final TaskAttachmentDao taskAttachmentDao;
  private final CaldavDao caldavDao;
  private final WorkManager workManager;
  private final TaskDao taskDao;
  private final UserActivityDao userActivityDao;
  private final Preferences preferences;
  private Context context;
  private int exportCount = 0;
  private ProgressDialog progressDialog;
  private Handler handler;
  private String latestSetVersionName;

  @Inject
  public TasksJsonExporter(
      TagDataDao tagDataDao,
      TaskDao taskDao,
      UserActivityDao userActivityDao,
      Preferences preferences,
      AlarmDao alarmDao,
      LocationDao locationDao,
      TagDao tagDao,
      GoogleTaskDao googleTaskDao,
      FilterDao filterDao,
      GoogleTaskListDao googleTaskListDao,
      TaskAttachmentDao taskAttachmentDao,
      CaldavDao caldavDao,
      WorkManager workManager,
      DriveInvoker driveInvoker) {
    this.tagDataDao = tagDataDao;
    this.taskDao = taskDao;
    this.userActivityDao = userActivityDao;
    this.preferences = preferences;
    this.alarmDao = alarmDao;
    this.locationDao = locationDao;
    this.tagDao = tagDao;
    this.googleTaskDao = googleTaskDao;
    this.filterDao = filterDao;
    this.googleTaskListDao = googleTaskListDao;
    this.taskAttachmentDao = taskAttachmentDao;
    this.caldavDao = caldavDao;
    this.workManager = workManager;
  }

  private static String getDateForExport() {
    return newDateTime().toString("yyMMdd-HHmm");
  }

  private void post(Runnable runnable) {
    if (handler != null) {
      handler.post(runnable);
    }
  }

  private void setProgress(final int taskNumber, final int total) {
    post(
        () -> {
          progressDialog.setMax(total);
          progressDialog.setProgress(taskNumber);
        });
  }

  public void exportTasks(
      final Context context,
      final ExportType exportType,
      @Nullable final ProgressDialog progressDialog) {
    this.context = context;
    this.exportCount = 0;
    this.latestSetVersionName = null;
    this.progressDialog = progressDialog;

    if (exportType == ExportType.EXPORT_TYPE_MANUAL) {
      handler = new Handler();
      new Thread(() -> runBackup(exportType)).start();
    } else {
      runBackup(exportType);
    }
  }

  private void runBackup(ExportType exportType) {
    try {
      String filename = getFileName(exportType);
      List<Task> tasks = taskDao.getAll();

      if (tasks.size() > 0) {
        String basename = Files.getNameWithoutExtension(filename);
        Uri uri =
            FileHelper.newFile(
                context, preferences.getBackupDirectory(), MIME, basename, EXTENSION);
        OutputStream os = context.getContentResolver().openOutputStream(uri);
        doTasksExport(os, tasks);
        os.close();
        workManager.scheduleDriveUpload(uri, exportType == ExportType.EXPORT_TYPE_SERVICE);
      }

      if (exportType == ExportType.EXPORT_TYPE_MANUAL) {
        onFinishExport(filename);
      }
    } catch (IOException e) {
      Timber.e(e);
    } finally {
      post(
          () -> {
            if (progressDialog != null
                && progressDialog.isShowing()
                && context instanceof Activity) {
              DialogUtilities.dismissDialog((Activity) context, progressDialog);
            }
          });
    }
  }

  private void doTasksExport(OutputStream os, List<Task> tasks) throws IOException {

    List<BackupContainer.TaskBackup> taskBackups = new ArrayList<>();

    for (Task task : tasks) {
      setProgress(taskBackups.size(), tasks.size());
      long taskId = task.getId();
      taskBackups.add(
          new BackupContainer.TaskBackup(
              task,
              alarmDao.getAlarms(taskId),
              locationDao.getGeofencesForTask(taskId),
              tagDao.getTagsForTask(taskId),
              googleTaskDao.getAllByTaskId(taskId),
              userActivityDao.getCommentsForTask(task.getUuid()),
              taskAttachmentDao.getAttachments(task.getUuid()),
              caldavDao.getTasks(taskId)));
    }

    Map<String, Object> data = new HashMap<>();
    data.put("version", BuildConfig.VERSION_CODE);
    data.put("timestamp", System.currentTimeMillis());
    data.put(
        "data",
        new BackupContainer(
            taskBackups,
            locationDao.getPlaces(),
            tagDataDao.getAll(),
            filterDao.getAll(),
            googleTaskListDao.getAccounts(),
            googleTaskListDao.getAllLists(),
            caldavDao.getAccounts(),
            caldavDao.getCalendars(),
            preferences.getPrefs(Integer.class),
            preferences.getPrefs(Long.class),
            preferences.getPrefs(String.class),
            preferences.getPrefs(Boolean.class)));

    OutputStreamWriter out = new OutputStreamWriter(os, UTF_8);
    Gson gson = BuildConfig.DEBUG ? new GsonBuilder().setPrettyPrinting().create() : new Gson();
    out.write(gson.toJson(data));
    out.close();
    exportCount = taskBackups.size();
  }

  private void onFinishExport(final String outputFile) {
    post(
        () -> {
          if (exportCount == 0) {
            Toast.makeText(
                    context, context.getString(R.string.export_toast_no_tasks), Toast.LENGTH_LONG)
                .show();
          } else {
            CharSequence text =
                String.format(
                    context.getString(R.string.export_toast),
                    context
                        .getResources()
                        .getQuantityString(R.plurals.Ntasks, exportCount, exportCount),
                    outputFile);
            Toast.makeText(context, text, Toast.LENGTH_LONG).show();
          }
        });
  }

  private String getFileName(ExportType type) {
    switch (type) {
      case EXPORT_TYPE_SERVICE:
        return String.format(BackupConstants.BACKUP_FILE_NAME, getDateForExport());
      case EXPORT_TYPE_MANUAL:
        return String.format(BackupConstants.EXPORT_FILE_NAME, getDateForExport());
      case EXPORT_TYPE_ON_UPGRADE:
        return String.format(BackupConstants.UPGRADE_FILE_NAME, latestSetVersionName);
      default:
        throw new UnsupportedOperationException("Unhandled export type");
    }
  }

  public enum ExportType {
    EXPORT_TYPE_SERVICE,
    EXPORT_TYPE_MANUAL,
    EXPORT_TYPE_ON_UPGRADE
  }
}

File: app/src/main/java/org/tasks/backup/TasksJsonExporter.kt
package org.tasks.backup

import android.app.Activity
import android.app.ProgressDialog
import android.app.backup.BackupManager
import android.content.Context
import android.net.Uri
import android.os.Handler
import com.google.common.io.Files
import com.todoroo.andlib.utility.DialogUtilities
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.tasks.BuildConfig
import org.tasks.R
import org.tasks.caldav.VtodoCache
import org.tasks.data.*
import org.tasks.data.dao.AlarmDao
import org.tasks.data.dao.CaldavDao
import org.tasks.data.dao.FilterDao
import org.tasks.data.dao.LocationDao
import org.tasks.data.dao.TagDao
import org.tasks.data.dao.TagDataDao
import org.tasks.data.dao.TaskAttachmentDao
import org.tasks.data.dao.TaskDao
import org.tasks.data.dao.TaskListMetadataDao
import org.tasks.data.dao.UserActivityDao
import org.tasks.date.DateTimeUtils.newDateTime
import org.tasks.extensions.Context.toast
import org.tasks.files.FileHelper
import org.tasks.jobs.WorkManager
import org.tasks.preferences.Preferences
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import timber.log.Timber
import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.io.Writer
import java.util.Set
import javax.inject.Inject

class TasksJsonExporter @Inject constructor(
    private val tagDataDao: TagDataDao,
    private val taskDao: TaskDao,
    private val userActivityDao: UserActivityDao,
    private val preferences: Preferences,
    private val alarmDao: AlarmDao,
    private val locationDao: LocationDao,
    private val tagDao: TagDao,
    private val filterDao: FilterDao,
    private val taskAttachmentDao: TaskAttachmentDao,
    private val caldavDao: CaldavDao,
    private val workManager: WorkManager,
    private val taskListMetadataDao: TaskListMetadataDao,
    private val vtodoCache: VtodoCache,
) {
    private var context: Context? = null
    private var exportCount = 0
    private var progressDialog: ProgressDialog? = null
    private var handler: Handler? = null

    private fun post(runnable: () -> Unit) = handler?.post(runnable)

    private fun setProgress(taskNumber: Int, total: Int) = post {
        progressDialog?.max = total
        progressDialog?.progress = taskNumber
    }

    suspend fun exportTasks(context: Context?, exportType: ExportType, progressDialog: ProgressDialog?) {
        this.context = context
        exportCount = 0
        this.progressDialog = progressDialog
        if (exportType == ExportType.EXPORT_TYPE_MANUAL) {
            handler = Handler()
        }
        runBackup(exportType)
    }

    private suspend fun runBackup(exportType: ExportType) {
        try {
            val filename = getFileName(exportType)
            val tasks = taskDao.getAllTaskIds()
            val file = File(String.format("%s/%s", context!!.filesDir, BackupConstants.INTERNAL_BACKUP))
            file.delete()
            file.createNewFile()
            val internalStorageBackup = Uri.fromFile(file)
            val os = context!!.contentResolver.openOutputStream(internalStorageBackup)
            doTasksExport(os, tasks)
            os!!.close()
            val externalStorageBackup = FileHelper.newFile(
                    context!!,
                    preferences.backupDirectory!!,
                    MIME,
                    Files.getNameWithoutExtension(filename),
                    EXTENSION)
            FileHelper.copyStream(context!!, internalStorageBackup, externalStorageBackup)
            workManager.scheduleDriveUpload(externalStorageBackup, exportType == ExportType.EXPORT_TYPE_SERVICE)
            BackupManager(context).dataChanged()
            if (exportType == ExportType.EXPORT_TYPE_MANUAL) {
                onFinishExport(filename)
            }
        } catch (e: IOException) {
            Timber.e(e)
        } finally {
            post {
                if (progressDialog != null && progressDialog!!.isShowing
                        && context is Activity) {
                    DialogUtilities.dismissDialog(context as Activity?, progressDialog)
                }
            }
        }
    }

    suspend fun doSettingsExport(os: OutputStream?) = withContext(Dispatchers.IO) {
        val writer = os!!.bufferedWriter()
        with (JsonWriter(writer)) {
            write("{")
            write("version", BuildConfig.VERSION_CODE)
            write("timestamp", currentTimeMillis())
            write("\"data\":{")
            writePreferences()
            write("}")
            write("}")
        }
        writer.flush()
    }

    @Throws(IOException::class)
    private suspend fun doTasksExport(os: OutputStream?, taskIds: List<Long>) = withContext(Dispatchers.IO) {
        val writer = os!!.bufferedWriter()
        with (JsonWriter(writer)) {
            write("{")
            write("version", BuildConfig.VERSION_CODE)
            write("timestamp", currentTimeMillis())
            write("\"data\":{")
            write("\"tasks\":[")
            taskIds.forEachIndexed { index, id ->
                setProgress(index, taskIds.size)
                write("{")
                write("task", taskDao.fetch(id)!!)
                write("alarms", alarmDao.getAlarms(id))
                write("geofences", locationDao.getGeofencesForTask(id))
                write("tags", tagDao.getTagsForTask(id))
                write("comments", userActivityDao.getComments(id))
                write("attachments", taskAttachmentDao.getAttachmentsForTask(id))
                val caldavTasks = caldavDao.getTasks(id)
                vtodoCache
                    .getVtodo(caldavTasks.firstOrNull { !it.isDeleted() })
                    ?.let { write("vtodo", it) }
                write("caldavTasks", caldavTasks, lastItem = true)
                write("}")
                if (index < taskIds.size - 1) write(",")
            }
            write("],")
            write("places", locationDao.getPlaces())
            write("tags", tagDataDao.getAll())
            write("filters", filterDao.getFilters())
            write("caldavAccounts", caldavDao.getAccounts())
            write("caldavCalendars", caldavDao.getCalendars())
            write("taskListMetadata", taskListMetadataDao.getAll())
            write("taskAttachments", taskAttachmentDao.getAttachments())
            writePreferences()
            write("}")
            write("}")
        }
        writer.close()
        os.close()
        exportCount = taskIds.size
    }

    private fun JsonWriter.writePreferences() {
        write("intPrefs", preferences.getPrefs(Integer::class.java))
        write("longPrefs", preferences.getPrefs(java.lang.Long::class.java))
        write("stringPrefs", preferences.getPrefs(String::class.java))
        write("boolPrefs", preferences.getPrefs(java.lang.Boolean::class.java))
        write("setPrefs", preferences.getPrefs(Set::class.java) as Map<String, Set<String>>, lastItem = true)
    }

    private fun onFinishExport(outputFile: String) = post {
        context?.toast(
            R.string.export_toast,
            context!!
                .resources
                .getQuantityString(R.plurals.Ntasks, exportCount, exportCount),
            outputFile
        )
    }


    private fun getFileName(type: ExportType): String =
        when (type) {
            ExportType.EXPORT_TYPE_SERVICE -> String.format(BackupConstants.BACKUP_FILE_NAME, dateForExport)
            ExportType.EXPORT_TYPE_MANUAL -> String.format(BackupConstants.EXPORT_FILE_NAME, dateForExport)
        }

    enum class ExportType {
        EXPORT_TYPE_SERVICE, EXPORT_TYPE_MANUAL
    }

    companion object {
        private const val MIME = "application/json"
        private const val EXTENSION = ".json"
        private val dateForExport: String
            get() = newDateTime().toString("yyyyMMdd'T'HHmm")

        class JsonWriter(val writer: Writer, val json: Json = Json) {
            fun write(data: String) = writer.write(data)

            inline fun <reified T> write(key: String, value: @Serializable T, lastItem: Boolean = false) where T : Any =
                writer.write("\"$key\":${json.encodeToString(value)}${if (lastItem) "" else ","}")
        }
    }
}